### PR TITLE
Add support for external images

### PR DESCRIPTION
**Pull request by fran-worley**
Adds basic support for images

**Limitations:**
- Images are external i.e. pictures accessed via URL, not stored within document
- only sizing is customisable

**Other changes:** 
In order for relationship referencing to work I have appended links with Href and images with Image. Couldn’t find a better way to reference the numbers correctly.

Beginning of fix for #27

### CODE REVIEW

**Comment by fran-worley** [comment]
Until I can find a way to save source files in `'/word/media/[image-filename]'` I can't add proper support for images. 

External images have the obvious drawback that they require an internet connection to load etc.

I will keep trying but if anyone has any bright ideas...

**Comment by fran-worley** [comment]
I have now got this working with internal images. 

To work correctly your image source must be the full url, not just the relative path and you must provide the image size in pixels or EMs.

**Size**
Size can be provided either via the style attribute:

```html
<p><img src="http://placehold.it/250x100.png" style="width: 250px; height: 100px"></p>
```
or via data attributes

```html
<p><img src="http://placehold.it/250x100.png" data-width="250px" data-height="100px"></p>
```
Should you provide both, data attributes take precedent.

**Filename**

The original filename of your image can either be inferred from the source:
```html
<p><img src="http://placehold.it/250x100.png" data-width="250px" data-height="100px"></p>
```
would give a filename of:  250x100png

or you can provide a value via data attributes:
```html
<p><img src="http://placehold.it/250x100.png" data-filename="what-a-lovely-image.png" data-width="250px" data-height="100px"></p>
```
would give a filename of:  what-a-lovely-image.png

This is useful when the source url doesn't include the file extension or contains special characters.

**Accessibility**
For accessibility you are recommended to provide titles for images via the alt attribute:

```html
<p><img src="http://placehold.it/250x100.png" alt="Fancy image description" style="height:100px; width:250px"></p>
```

**To do:**

- Find a better way to set the relationship ids for external links and all images.
- Find a better way to add required content types into file.
- Add support for customising position and text-wrap settings
- Enable the domain to be provided via gem config & then accept relative source paths

**Comment by anitsirc** [comment]
Great stuff @fran-worley! I'll review it during the week. Will you work on the todo or it was more like an informative todo?

**Comment by fran-worley** [comment]
I'm planning to address them with further PRs. However I would like to discuss points 2 &4 as they potentially impact the core of the gem. Also, I couldn't see any testing for document.rb which seems fairly critical give that the processing happens here.

File: lib/htmltoword/xslt/image_functions.xslt [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+    Returns the unqualified dimension from a length specification copied from:
+  http://docbook.sourceforge.net/release/xsl/1.76.1/doc/lib/length-magnitude.html
+   -->
+
+  <xsl:template name="length-magnitude">
+    <xsl:param name="length" select="'0pt'"></xsl:param>
+
+    <xsl:choose>
+      <xsl:when test="string-length($length) = 0"></xsl:when>
+      <xsl:when test="substring($length,1,1) = '0'                     or substring($length,1,1) = '1'                     or substring($length,1,1) = '2'                     or substring($length,1,1) = '3'                     or substring($length,1,1) = '4'                     or substring($length,1,1) = '5'                     or substring($length,1,1) = '6'                     or substring($length,1,1) = '7'                     or substring($length,1,1) = '8'                     or substring($length,1,1) = '9'                     or substring($length,1,1) = '.'">
```
Comment (anitsirc) [inline lib/htmltoword/xslt/image_functions.xslt]: is there a need for so many blank spaces?

Comment (fran-worley) [inline lib/htmltoword/xslt/image_functions.xslt]: No need, just helped to break up the numerous conditions though I appreciate they aren't aligned here. 

Shall I line them up? I'm loathed to take out all of the spaces as then it will be impossible to read...

**Comment by anitsirc** [comment]
@fran-worley I think there should be default dimensions when it's not defined in the image, or ignore and don't add it if not present. We shouldn't punish the users having images, generating empty or corrupted files, just because they haven't define a width and height.

File: lib/htmltoword/xslt/base.xslt [hunk 2]
```diff
@@ -204,6 +204,12 @@
         <xsl:when test="self::a[starts-with(@href, 'http://') or starts-with(@href, 'https://')]">
           <xsl:call-template name="link" />
         </xsl:when>
+        <xsl:when test="self::img">
```
Comment (anitsirc) [inline lib/htmltoword/xslt/base.xslt]: I think this is not working, also if we support `<a...></a>` this should also be in the link template on `links.xslt`

File: lib/htmltoword/xslt/images.xslt [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+                xmlns:func="http://exslt.org/functions"
+                xmlns:fn="http://www.w3.org/2005/xpath-functions"
+                mc:Ignorable="w14 w15 wp14 a14"
+                version="1.0"
+                exclude-result-prefixes="java msxsl ext w o v WX aml w10"
+                extension-element-prefixes="func">
+
+  <xsl:include href="./image_functions.xslt"/>
+
+  <xsl:template match="img" name="image">
```
Comment (anitsirc) [inline lib/htmltoword/xslt/images.xslt]: Can you change the match to be `match="img|body/img"` ? 

There's a bug caused by https://github.com/karnov/htmltoword/blob/master/lib/htmltoword/xslt/base.xslt#L44 which makes markup like the following one to fail. 

```html
<body>
  <img src="https://assets-cdn.github.com/images/modules/logos_page/Octocat.png" style="width: 250px; height: 100px">
</body>
```

I'm still trying to figure out why we needed that line...

File: spec/xslt_images_spec.rb [hunk 4]
```diff
@@ -0,0 +1,318 @@
...
+    <Relationship Type="http://schemas.openxmlformats.org/officeDocument/2006/relationships/hyperlink" Target="http://somelink.com" TargetMode="External" Id="rId8"/>
+    <Relationship Type="http://schemas.openxmlformats.org/officeDocument/2006/relationships/image" Target="http://placehold.it/350x150" TargetMode="External" Id="rId9"/>
+    <Relationship Type="http://schemas.openxmlformats.org/officeDocument/2006/relationships/hyperlink" Target="http://someotherlink2.com" TargetMode="External" Id="rId10"/>
+    <Relationship Type="http://schemas.openxmlformats.org/officeDocument/2006/relationships/image" Target="media/image2.png" Id="rId11"/>
+    <Relationship Type="http://schemas.openxmlformats.org/officeDocument/2006/relationships/image" Target="media/image3.jpg" Id="rId12"/>
+  </Relationships>
+    EOL
+    compare_relations_xml(html, expected_relations_xml)
+  end
+end
```
Comment (anitsirc) [inline spec/xslt_images_spec.rb]: Add newline at the end of the file

File: lib/htmltoword/xslt/image_functions.xslt [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+          <xsl:value-of select="translate(str:tokenize(substring-after($style, concat($type,':')), ';')[1],'&#x20;&#x9;&#xD;&#xA;','')" />
+        </xsl:when>
+        <xsl:otherwise>1</xsl:otherwise>
+      </xsl:choose>
+    </xsl:variable>
+    <xsl:call-template name="length-in-emus">
+      <xsl:with-param name="length" select="$size"/>
+    </xsl:call-template>
+  </xsl:template>
+</xsl:stylesheet>
```
Comment (anitsirc) [inline lib/htmltoword/xslt/image_functions.xslt]: Add newline at the end of the file

File: lib/htmltoword/xslt/images.xslt [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+    <xsl:attribute name="cy">
+      <xsl:call-template name="image-dimention">
+        <xsl:with-param name="style" select="@style" />
+        <xsl:with-param name="data-value" select="@data-height" />
+        <xsl:with-param name="type" select="'height'" />
+      </xsl:call-template>
+    </xsl:attribute>
+  </xsl:template>
+
+</xsl:stylesheet>
```
Comment (anitsirc) [inline lib/htmltoword/xslt/images.xslt]: Add newline at the end of the file

File: lib/htmltoword/document.rb [hunk 7]
```diff
@@ -61,10 +66,21 @@ def generate
               out.write(source)
             elsif @replaceable_files[entry.name]
               out.write(@replaceable_files[entry.name])
+            elsif entry.name == Document.content_types_xml_file
```
Comment (anitsirc) [inline lib/htmltoword/document.rb]: Not a big fan of this, specially if it might affect the custom templates users might use (affecting a PartName or something). What about maybe using a transformation to add the ContentTypes that are actually in use?

**Comment by fran-worley** [comment]
Thanks for the feedback. I have a couple of questions...

**Default Image size**
In my opinion a default image size isn't going to work as rendering all images at one size will make your document look awful. Either we don't show images at all without a size or we include  a library like [FastImage] (https://github.com/sdsykes/fastimage) to calculate the size from source if the user doesn't specify one.

**Supporting images in links**
This is more complicated as you don't appear to be able to nest relationships. I've pulled the xml that word generates when including images in links and it doesn't use relationships or hyperlink tags at all:

```xml
  <w:p w14:paraId="0533D16D" w14:textId="77777777" w:rsidR="00E55D2E" w:rsidRPr="000B537C" w:rsidRDefault="000B537C">
    
    <w:pPr>
      <w:rPr>
        <w:rStyle w:val="Hyperlink"/>
      </w:rPr>
    </w:pPr>

    <w:r>
      <w:rPr>
        <w:noProof/>
        <w:lang w:eastAsia="en-US"/>
      </w:rPr>
      <w:drawing><xsl:comment>Some lovely image xml</xsl:comment></w:drawing>
    </w:r>

    <w:r>
      <w:fldChar w:fldCharType="begin"/>
    </w:r>

    <w:r>
      <w:instrText xml:space="preserve"> HYPERLINK "http://www.example.com/" </w:instrText>
    </w:r>

    <w:r>
      <w:fldChar w:fldCharType="separate"/>
    </w:r>

    <w:r w:rsidRPr="000B537C">
      <w:rPr>
        <w:rStyle w:val="Hyperlink"/>
      </w:rPr>
      <w:t>Link Text</w:t>
    </w:r>
  </w:p>
```

I'm not sure why there is a difference in markup used by word and I can't find any documentation of this markup in the openoffice docs. If we want to support images then I'll need to rewrite the xml for links.

**content types**
This was on my list as I didn't want to have to replace the entire file, but the files are corrupted if they don't contain the correct mimetypes.

We could open the file and inject the relevant `default` tags at the end of the `Types` tag as it doesn't appear to matter what order they are in.

Not tested yet but thinking something like this...
```ruby
#replace current content_type code in #generate with this...
if entry.name == Document.content_types_xml_file
 inject_image_content_types(entry) if @image_files.size > 0
end


#add a method to document.rb to inject the required content_types into the file...
def inject_image_content_types(file)
  doc = Nokogiri::XML(File.open(file)) 
  #get a list of all extensions currently in content_types file
  existing_exts = doc.xpath("/Default").map { |node| node.attribute("Extension") }.compact 
  
  #get a list of extensions we need for our images
  required_exts = @image_files.map{ |i| i[:ext] }

  #workout which required extensions are missing from the content_types file
  missing_exts = required_exts - (existing_exts & required_exts)

  #inject missing extensions into document
  missing_exts.each do |ext|
    doc.at_css("Types").add_child( "<Default Extension='#{ext}' ContentType='image/#{ext}'/>")
  end
end
```


@anitsirc Thoughts??

File: lib/htmltoword/document.rb (hunk 7 above)
Comment (fran-worley) [inline lib/htmltoword/document.rb]: I've changed this with my latest commit. We now read the contents of the content_types file and inject the required tags based on the image extensions included in the document.

**Comment by fran-worley** [comment]
@anitsirc Any chance you can have a look at where I've got to. I'd love to get this merged soon...

**Comment by jiek85** [comment]
Unrecognized unit of measure: .?

**Comment by fran-worley** [comment]
Currently you must provide a width and height in pixels or ems. You can do so either via the style or data attributes (data takes president should both be found) 

If you don't provide a size or the size is in another unit (e.g percentage) you'll get an error.

**Comment by tilsammans** [comment]
@anitsirc any chance this can be merged? Would be lovely to have!

**Comment by mojarra** [comment]
@anitsirc any problems to merge this? It would be awesome

**Comment by tilsammans** [comment]
@karnov @nickfrandsen ping ... anyone please

**Comment by stats** [comment]
Hi, jumping into this conversation a little late.

On windows 7 64-bit ruby 2.3.1p112. Make a simple project:

**gemfile**
```
source 'https://rubygems.org'
gem 'htmltoword', git: 'https://github.com/fran-worley/htmltoword', branch: 'images-external'
```

testhtmltoword.rb
```
require 'htmltoword'

my_html = '<html><head></head><body><p>Hello</p></body></html>'
document = Htmltoword::Document.create(my_html)
file = Htmltoword::Document.create_and_save(my_html, 'test.docx')
```

running *bundle exec ruby testhtmltoword.rb* results in a *test.docx* file that cannot open in Word 2013.

The error is: 
```
We're sorry. We can't open test.docx because we found a problem with its contents.

Details: The file is corrupt and cannot be opened.
```

Using the default htmltoword gem (0.5.1) the same test file produces a Word document that can be opened. 

I used WinMerge to look for differences and the only difference I found was in the **[Content_Types].xml** file, in the version from your branch I found the following, while from the 0.5.1 code there is nothing:
```
<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
<Types xmlns="http://schemas.openxmlformats.org/package/2006/content-types">
  <Default Extension="xml" ContentType="application/xml"/>
  <Default Extension="rels" ContentType="application/vnd.openxmlformats-package.relationships+xml"/>
  <Default Extension="jpeg" ContentType="image/jpeg"/>
  <Override PartName="/word/document.xml" ContentType="application/vnd.openxmlformats-officedocument.wordprocessingml.document.main+xml"/>
  <Override PartName="/word/numbering.xml" ContentType="application/vnd.openxmlformats-officedocument.wordprocessingml.numbering+xml"/>
  <Override PartName="/word/styles.xml" ContentType="application/vnd.openxmlformats-officedocument.wordprocessingml.styles+xml"/>
  <Override PartName="/word/stylesWithEffects.xml" ContentType="application/vnd.ms-word.stylesWithEffects+xml"/>
  <Override PartName="/word/settings.xml" ContentType="application/vnd.openxmlformats-officedocument.wordprocessingml.settings+xml"/>
  <Override PartName="/word/webSettings.xml" ContentType="application/vnd.openxmlformats-officedocument.wordprocessingml.webSettings+xml"/>
  <Override PartName="/word/fontTable.xml" ContentType="application/vnd.openxmlformats-officedocument.wordprocessingml.fontTable+xml"/>
  <Override PartName="/word/theme/theme1.xml" ContentType="application/vnd.openxmlformats-officedocument.theme+xml"/>
  <Override PartName="/docProps/core.xml" ContentType="application/vnd.openxmlformats-package.core-properties+xml"/>
  <Override PartName="/docProps/app.xml" ContentType="application/vnd.openxmlformats-officedocument.extended-properties+xml"/>
</Types>
```

Not sure why this would cause an issue.

**Comment by fran-worley** [comment]
@stats can you attach your word document?

**Comment by stats** [comment]
test-0.6.0.docx is your branch.
test-0.5.1.docx is from the karnov/htmltoword master

For some reason your branch is not including the contents for [Content_Types].xml.

After some additional testing I think there there may be a problem if no image file is included in the document. In that case you will not get any content in the [Content_Types].xml file.

[test-0.6.0.docx](https://github.com/karnov/htmltoword/files/506600/test-0.6.0.docx)
[test-0.5.1.docx](https://github.com/karnov/htmltoword/files/506601/test-0.5.1.docx)

**Comment by fran-worley** [comment]
Good spot @stats that should now be fixed via https://github.com/karnov/htmltoword/pull/44/commits/a12c1e45d6c0c04ea084e5af2ff0b08ba3e33ec4

**Comment by stats** [comment]
Awesome. Thank you very much.

**Comment by fran-worley** [comment]
Do note the caviats in this branch.

1) images inside links do not render
2) all images must have their size width and height in pixels or ems (not %)
3) because images have to be downloaded and saved into the document it's not the quickest and you'll probably want to generate your docs in a background process.

**Comment by dochoaj** [comment]
@karnov are you planning to merge this?

**Comment by filipkis** [comment]
There is a small issue with current image naming functions in XSLT, that is if the source file was `original.name.png` and no data-filename was provided the file will be stored as `word/media/image1.png`, but in the relations file it will be referred to as: 

    <Relationship Type="http://schemas.openxmlformats.org/officeDocument/2006/relationships/image"
      Target="media/image1.name.png" 
      Id="rId8"/>

In other words the transformation takes `name.png` as extension and not just `png`.

**Comment by lukelex** [comment]
Hi all,

The development around this project has been slow from our side mostly because it already fits all of our use cases. We're willing to reignite the work here and on board anyone that has already contributed to this project. 

@fran-worley Sorry for the **super** late follow up on this.

**Comment by fran-worley** [comment]
@filipkis Good spot, I'll take a look at this and add to this PR.

@lukelex Good to here that you're looking at taking this on further if I can do anything to help get this branch merged let me know.

**Comment by eviofragoso** [comment]
Hello,
I'm trying to get images to work in the doc.

my tag is going like this: "<p><img alt=\"\" src=\"/ckeditor_assets/pictures/1/content_equi.jpg\" style=\"height:334px; width:375px\". 
It seems to be in the proper syntax to the gem to work, but i'ts ignored.

**Comment by kreintjes** [comment]
@fran-worley @anitsirc Please see https://github.com/karnov/htmltoword/issues/71#issuecomment-398437050. The images won't show up in the Word files I create. Both the `data-external` images as well as the internal images simply won't show up. I can generate the Word document (don't get any errors), but the images aren't there.

Your sample code: `<p>Fancy image description</p>` doesn't work either.

Could you help me with this?
